kratos/utilities/variable_utils.cpp

```diff
@@ -486,4 +487,94 @@ void VariableUtils::AuxiliaryAtomicAdd(
         rSumValue[2] += rPrivateValue[2];
 }
 
+template <>
+ModelPart::ElementsContainerType& VariableUtils::GetContainer<ModelPart::ElementsContainerType>(ModelPart& rModelPart)
+{
+    return rModelPart.Elements();
+}
+
+template <>
+ModelPart::ConditionsContainerType& VariableUtils::GetContainer<ModelPart::ConditionsContainerType>(ModelPart& rModelPart)
+{
+    return rModelPart.Conditions();
+}
+
+template <class TDataType, class TContainerType, class TWeightDataType>
+void VariableUtils::DistributeVariable(
+    ModelPart& rModelPart,
+    const Variable<TDataType>& rVariable,
+    const Variable<TWeightDataType>& rWeightVariable,
+    const bool IsInverseWeightProvided)
+{
+    KRATOS_TRY
+
+    SetNonHistoricalVariableToZero(rVariable, rModelPart.Nodes());
+
+    auto& r_entities = GetContainer<TContainerType>(rModelPart);
+    const int number_of_conditions = r_entities.size();
+
+    const std::function<double(const Node<3>&)>& r_direct_method =
+        [rWeightVariable](const Node<3>& rNode) -> double {
+            return rNode.GetValue(rWeightVariable);
+        };
+
+    const std::function<double(const Node<3>&)>& r_inverse_method =
+        [rWeightVariable](const Node<3>& rNode) -> double {
+            return 1.0 / rNode.GetValue(rWeightVariable);
+        };
+
+    const std::function<double(const Node<3>&)>& r_weight_method =
+        (IsInverseWeightProvided) ? r_inverse_method : r_direct_method;
+
+#pragma omp parallel for
+    for (int i_condition = 0; i_condition < number_of_conditions; ++i_condition)
+    {
+        auto it_entity = r_entities.begin() + i_condition;
+        auto& r_geometry = it_entity->GetGeometry();
+
+        const auto& r_value = it_entity->GetValue(rVariable);
+        for (int i_node = 0; i_node < static_cast<int>(r_geometry.PointsNumber()); ++i_node)
+        {
+            auto& r_node = r_geometry[i_node];
+
+            KRATOS_DEBUG_ERROR_IF(!r_node.Has(rWeightVariable))
+                << "Non-historical nodal " << rWeightVariable.Name() << " at "
+                << r_node << " is not initialized in " << rModelPart.Name()
+                << ". Please initialize it first.";
+
+            const auto& r_current_value = r_node.GetValue(rVariable);
+            const double weight = r_weight_method(r_node);
+
+            r_node.SetLock();
+            r_node.SetValue(rVariable, r_current_value + r_value * weight);
```

Comment:
```suggestion
            const double weight = r_weight_method(r_node);

            r_node.SetLock();
            r_node.GetValue(rVariable) += r_value * weight;
```

Reply:
I didn't know with Locks, you can use GetValue in OMP loops, I was trying to use it previously, but in full debug it failed. Thanks :)